pallets/grants/src/migrations.rs

```diff
@@ -0,0 +1,199 @@
+	use sp_std::vec::Vec;
+
+	pub struct MigrateToBoundedVestingSchedules<T>(PhantomData<T>);
+	impl<T: Config> OnRuntimeUpgrade for MigrateToBoundedVestingSchedules<T> {
+		fn on_runtime_upgrade() -> Weight {
+			let current = Pallet::<T>::current_storage_version();
+			let onchain = Pallet::<T>::on_chain_storage_version();
+
+			log::info!(
+				"on_runtime_upgrade[{:#?}]=> Running migration with current storage version {:?} / onchain {:?}",
+				line!(),
+				current,
+				onchain
+			);
+
+			if current == 1 && onchain == 0 {
+				let pallet_prefix: &[u8] = b"Vesting";
+				let storage_item_prefix: &[u8] = b"VestingSchedules";
+
+				// Check number of entries, and set it aside in temp storage
+				let stored_data: Vec<_> =
+					storage_key_iter::<T::AccountId, Vec<VestingScheduleOf<T>>, Blake2_128Concat>(
+						pallet_prefix,
+						storage_item_prefix,
+					)
+					.collect();
+
+				let mut translated = 0u64;
+				let mut max_schedules: usize = 0;
+
+				assert!(stored_data.len() > 0);
+
+				// Write to the new storage with removed and added fields
+				for (account, old_vesting) in stored_data {
+					translated.saturating_inc();
+					max_schedules = max_schedules.max(old_vesting.len());
+					let new_vesting: BoundedVec<VestingScheduleOf<T>, T::MaxSchedule> = old_vesting
+						.clone()
+						.try_into()
+						.map_err(|err| {
+							log::error!(
+								"on_runtime_upgrade[{:#?}]=> Schedule length :: {:#?} Max :: {:#?}",
+								line!(),
+								old_vesting.len(),
+								T::MaxSchedule::get()
+							);
+							err
+						})
+						.expect("Could be boundedvec Overflow");
+
+					<VestingSchedules<T>>::insert(account, new_vesting);
+				}
+
+				current.put::<Pallet<T>>();
+
+				log::info!(
+					"on_runtime_upgrade[{:#?}]=> Upgraded {} schedules, Max {}, storage to version {:?}",
+					line!(),
+					translated,
+					max_schedules,
+					current
+				);
+
+				T::DbWeight::get().reads_writes(translated + 1, translated + 1)
+			} else {
+				log::info!(
+					"on_runtime_upgrade[{:#?}]=> Migration did not executed. This probably should be removed",
+					line!(),
+				);
+				T::DbWeight::get().reads(1)
+			}
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			let current = Pallet::<T>::current_storage_version();
+			let onchain = Pallet::<T>::on_chain_storage_version();
+
+			log::info!(
+				"pre_upgrade[{:#?}]=> with current storage version {:?} / onchain {:?}",
+				line!(),
+				current,
+				onchain
+			);
+
+			if current == 1 && onchain == 0 {
+				let pallet_prefix: &[u8] = b"Vesting";
+				let storage_item_prefix: &[u8] = b"VestingSchedules";
+
+				// Check number of entries, and set it aside in temp storage
+				let stored_data: Vec<_> =
+					storage_key_iter::<T::AccountId, Vec<VestingScheduleOf<T>>, Blake2_128Concat>(
+						pallet_prefix,
+						storage_item_prefix,
+					)
+					.collect();
```

Comment:
I would suggest that after this stage we go through all the stored_date and check none of the accounts have grants more than the MaxSchedule. If we can show that in `pre-upgrade`, we can be confident `on_runtime_upgrade` is not going to fail. It makes on_runtime_upgrade safer because a failure would be flagged before making any changes.

Reply:
Done